### PR TITLE
Add Support for `DateOnly` & `TimeOnly` for `SupplyParameterFromQuery`

### LINKED ISSUE #35525: Support DateOnly and TimeOnly as parsing options for [SupplyParameterFromQuery]
### Is your feature request related to a problem? Please describe.
I want to use the `DateOnly` or `TimeOnly` types to parse querystring parameters but when I try to use either of those types I get:
```NotSupportedException: Querystring values cannot be parsed as type 'System.DateOnly'.```.
Seeing as `DateTime` is supported it would be great if we can add these types as well.

### Describe the solution you'd like
Support for the following code:
```
[Parameter, SupplyParameterFromQuery]
public DateOnly CurrentDate { get; set; }
[Parameter, SupplyParameterFromQuery]
public TimeOnly CurrentTime { get; set; }
```
I'd be happy to submit a PR if we want to support these scenarios.

### LINKED ISSUE #35567: `{Date, Time}Only` Querystring building APIs for Blazor
## Background and Motivation

This is an extension of https://github.com/dotnet/aspnetcore/issues/34115 to support `DateOnly` and `TimeOnly`, given we already support `DateTime`. 

## Proposed API

```diff
namespace Microsoft.AspNetCore.Components
{
    public static class NavigationManagerExtensions
    {
+       public static string GetUriWithQueryParameter(this NavigationManager navigationManager, string name, DateOnly value)
+       public static string GetUriWithQueryParameter(this NavigationManager navigationManager, string name, DateOnly? value)
+       public static string GetUriWithQueryParameter(this NavigationManager navigationManager, string name, TimeOnly value)
+       public static string GetUriWithQueryParameter(this NavigationManager navigationManager, string name, TimeOnly? value)
    }
}
```

## Usage Examples

``` C#
[Parameter, SupplyParameterFromQuery]
public DateOnly OnlyTheDate { get; set; }

[Parameter, SupplyParameterFromQuery]
public TimeOnly OnlyTheTime { get; set; }
```

## Alternative Designs

N/A, extension of existing API.

## Risks

Concept count / complexity. Other than that, it's a fairly straightforward change that a [user has requested](https://github.com/dotnet/aspnetcore/issues/35525) already.

### PR DESCRIPTION
**Pull request by TanayParikh**
Fixes: https://github.com/dotnet/aspnetcore/issues/35525
API Proposal Fixes: https://github.com/dotnet/aspnetcore/issues/35567

### CODE REVIEW

File: src/Components/test/testassets/BasicTestApp/RouterTest/WithQueryParameters.razor [hunk 1]
```diff
@@ -10,7 +12,6 @@
 <p>
     Links:
     <a href="WithQueryParameters/@FirstName?intvalue=123">With IntValue</a> |
-    <a href="WithQueryParameters/@FirstName?intvalue=123&NullableDateTimeValue=@(new DateTime(2000, 1, 2, 3, 4, 5, 6).ToString("u"))">With NullableDateTimeValue</a> |
```
Comment (TanayParikh) [inline src/Components/test/testassets/BasicTestApp/RouterTest/WithQueryParameters.razor:13]: This wasn't being utilized anywhere.

**Review by MackinnonBuck — APPROVED** [review]
Looks great to me!

**Review by pranavkm — APPROVED** [review]

File: src/Components/Components/src/Routing/RouteConstraint.cs [hunk 2]
```diff
@@ -27,6 +27,8 @@ public static UrlValueConstraint Parse(string template, string segment, string c
         {
             "bool" => typeof(bool),
             "datetime" => typeof(DateTime),
+            "dateonly" => typeof(DateOnly),
```
Comment (TanayParikh) [inline src/Components/Components/src/Routing/RouteConstraint.cs:30]: Remove this, will be done in https://github.com/dotnet/aspnetcore/issues/34591

**Comment by TanayParikh** [comment]
*do not merge, just resolving some e2e testing issues locally*

File: src/Components/Components/src/Routing/RouteConstraint.cs [hunk 3]
```diff
@@ -27,8 +27,6 @@ public static UrlValueConstraint Parse(string template, string segment, string c
         {
             "bool" => typeof(bool),
             "datetime" => typeof(DateTime),
-            "dateonly" => typeof(DateOnly),
-            "timeonly" => typeof(TimeOnly),
```
Comment (TanayParikh) [inline src/Components/Components/src/Routing/RouteConstraint.cs]: Removed this, will be done in https://github.com/dotnet/aspnetcore/issues/34591

File: src/Components/test/E2ETest/Tests/RoutingTest.cs [hunk 4]
```diff
@@ -808,7 +805,7 @@ public void CanArriveAtQueryStringPageWithDateTimeQuery()
             var dateTime = new DateTime(2000, 1, 2, 3, 4, 5, 6);
             var dateOnly = new DateOnly(2000, 1, 2);
             var timeOnly = new TimeOnly(3, 4, 5, 6);
-            SetUrlViaPushState($"/WithQueryParameters/Abc?&NullableDateTimeValue={dateTime.ToString("u")}&NullableDateOnlyValue={dateOnly.ToString("u")}&NullableTimeOnlyValue={timeOnly.ToString("u")}");
+            SetUrlViaPushState($"/WithQueryParameters/Abc?NullableDateTimeValue=2000-01-02%2003:04:05&NullableDateOnlyValue=2000-01-02&NullableTimeOnlyValue=03:04:05");
```
Comment (TanayParikh) [inline src/Components/test/E2ETest/Tests/RoutingTest.cs]: Constructing URL via interpolation was causing issues (due to localization / date cultures). Testing with a hard-coded string for now.

**Review by TanayParikh — COMMENTED** [review]
Setting this to auto-merge now, please let me know if there are any concerns.

**Review by pranavkm — APPROVED** [review]

**Pull request merged**